include/crocoddyl/core/action-base.hpp

```diff
@@ -217,9 +248,15 @@ class ActionModelAbstractTpl {
    */
   virtual void print(std::ostream& os) const;
 
+ private:
+  std::size_t ng_internal_;  //!< Internal object for storing the number of inequatility constraints
+  std::size_t nh_internal_;  //!< Internal object for storing the number of equatility constraints
+
  protected:
   std::size_t nu_;                          //!< Control dimension
   std::size_t nr_;                          //!< Dimension of the cost residual
+  std::size_t* ng_;                         //!< Number of inequality constraints
```

Comment:
> Ok, I reviewed the code. Seems good in general. The "shareMemory" design is kept from cost-sum, which is good.
> 
> There is three major flaws to possibly discuss:
> 
> * the strange design pattern you used to store the ng and nh counters.

This design is needed for MPC applications when the dimension of constraints (e.g., contact constraints) changes at each MPC iteration. Again, I introduced this strategy needed in our inverse-dynamics MPC. Do you see a better alternative here? (e.g., smart pointers) 

> * the choice of allowing various constraint types (==, <=, <=<=) for a single residual. I dont think it is a good idea as it would be much less efficient to copy the memory from the residual model derivatives to the constraint. And I dont see much use of this in practice, do you?

This is indeed a good point. We could create different ConstraintResidual classes, one generic as proposed here. And other ones tailored to the cases ==, < and >. Let me know if you have alternative suggestions here as well. (I could implement these solutions in this PR or later; depending on how critical is to have it now)


> * you are evaluating double bounds as two rows in the inequality vector and derivatives. This is not efficient and many solvers would make use of the information.

It seems a reasonable choice to not tighten the bounds. I will handle it soon.